### ISSUE TITLE
SS4 upgrade discussion

### ISSUE THREAD

**Issue by nspyke** [unclecheese/silverstripe-dashboard#87]
Hey @unclecheese 
As I discussed with you yesterday at StripeCon, I plan on upgrading this module to SS4.
I'd like to open a discussion on the tasks needed to complete the upgrade.

- Run upgrader tool
- Namespacing
- Check and upgrade data model relationships
- ?

You mentioned that you'd like to remove some of the dashboards. I remember that you mentioned the Google Analytics dashboard. What were the other ones?

**Comment by mlewis-everley** [comment]
I may be able to help out (although I am pretty stacked on workload at the moment).

From my perspective, it would be nice if dashboard could loose the CMS dependency, especially as SS4 has decoupled so much functionality into separate modules. SS is a very nice framework and adding dashboards to custom apps would be (in my opinion) a very nice feature.

I feel that the following panels would be better in their own module (not part of core):

1. Google Analytics.
2. Blog Entry.
3. Maybe weather?

With regards to namespacing, I am guessing it would be: `namespace UncleCheese\Dashboard\xxx`

Finally, would you keep data model table names the same? If so, upgrading should be minimal really, if not then I am guessing a custom migration task will need creating (that would also cover point 1).

The only other thing that would need to be considered is if any of the JS needs updating. So far, in my experience, old Entwine code seems to work pretty well in 4, but there may be issues.

**Comment by adrian-stein** [comment]
Did work ever start on an SS4 branch?

**Comment by nspyke** [comment]
No. I ended up dropping this module from my upgrade project. I don't see myself working on it any time soon.

**Comment by nspyke** [comment]
No. I ended up dropping this module from my upgrade project. I don't see myself working on it any time soon.

**Comment by mlewis-everley** [comment]
I may get a chance to do some work on this soon (no guarantees though). I am going to need it to support some other modules I am upgrading to 4, but it really depends on if I can fit it in around other projects :-s

**Comment by Taitava** [comment]
How about creating the first SS4 compatible version without any of the default dashboards? Those could be upgraded and readded one by one to later minor versions.

That would satisfy projects that only use custom dashboards.

**Comment by Taitava** [comment]
I've started upgrading this the way I described in the above message (removed all default dashboards, those can be upgraded and brought back later). I'm not sure if I'm able to get it finished, but at least I'm trying. I'll push to this branch when I have it in a better shape: https://github.com/Taitava/silverstripe-dashboard/tree/ss4

So check that branch if there's no update in this discussion. If I'm not able to finish this, I'll at least push a WIP commit so someone else may get perhaps a little bit easier starting point.

**Comment by Taitava** [comment]
I'm quite close to completing the upgrade. Currently I need to test the changes with one or two of my projects. I'm not able to do comprehensive testing, but at least some. I will create a pull request after I'm done testing. In the meanwhile, I think others can test my branch too if someone is in a hurry or wants to help.

**Comment by 3Dgoo** [comment]
My work mates and I have developed an alternate dashboard module that works with SilverStripe 4: 
https://github.com/plastyk/silverstripe-dashboard

If anyone is still waiting for an SS4 dashboard module, feel free to give it a go. Hopefully it is useful.
